plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/DockerInstanceTest.java

```diff
@@ -0,0 +1,239 @@
+import org.eclipse.che.api.machine.server.exception.MachineException;
+import org.eclipse.che.api.machine.server.model.impl.LimitsImpl;
+import org.eclipse.che.api.machine.server.model.impl.MachineConfigImpl;
+import org.eclipse.che.api.machine.server.model.impl.MachineImpl;
+import org.eclipse.che.api.machine.server.model.impl.MachineSourceImpl;
+import org.eclipse.che.api.machine.server.spi.InstanceKey;
+import org.eclipse.che.plugin.docker.client.DockerConnector;
+import org.eclipse.che.plugin.docker.client.Exec;
+import org.eclipse.che.plugin.docker.client.LogMessage;
+import org.eclipse.che.plugin.docker.client.MessageProcessor;
+import org.eclipse.che.plugin.docker.client.ProgressMonitor;
+import org.eclipse.che.plugin.docker.client.params.CommitParams;
+import org.eclipse.che.plugin.docker.client.params.PushParams;
+import org.eclipse.che.plugin.docker.machine.node.DockerNode;
+import org.mockito.Mock;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.anyVararg;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Tests for {@link DockerInstance}.
+ *
+ * @author Anton Korneta
+ */
+@Listeners(MockitoTestNGListener.class)
+public class DockerInstanceTest {
+    private static final String        FILE_PATH    = "/tmp";
+    private static final String        CONTAINER    = "container144";
+    private static final String        OWNER        = "owner12";
+    private static final String        IMAGE        = "iamage12";
+    private static final String        MACHINE_ID   = "machine12";
+    private static final String        WORKSPACE_ID = "workspace12";
+    private static final String        NAME         = "suse-jdk";
+    private static final String        TYPE         = "docker";
+    private static final String        REGISTRY     = "registry";
+    private static final String        REPOSITORY   = "eclipse-che";
+    private static final String        TAG          = "latest";
+    private static final MachineStatus STATUS       = MachineStatus.RUNNING;
+
+    @Mock
+    private LogMessage                 logMessageMock;
+    @Mock
+    private Exec                       execMock;
+    @Mock
+    private DockerConnector            dockerConnectorMock;
+    @Mock
+    private DockerInstanceStopDetector dockerInstanceStopDetectorMock;
+
+    private DockerInstance dockerInstance;
+
+    @BeforeMethod
+    public void setUp() throws IOException {
+        dockerInstance = getDockerInstance();
```

Comment:
my point are
1. when a new parameter is added in constructor, then we need to adjust constructor
so if I create object in each method, I need to add this new dummy object in every calls (adding many lines) even if I don't need this object in the test method.
2. By only testing the class matching <ClassName>Test.class and mocking the other dependencies we're sure to test the behavior of this object and to reach pretty close 100% of code coverage while by using some real objects.
3. I don't see the point of "changing complex object constructors" because maybe if you follow TDD approach you would have modify the tests before modifying the code. about troubleshooting, if a required parameter is not injected then instance won't be created and you'll get error